File: domains/flight-shared-types/tests/integration/go/memory_types_test.go
// Go Integration Tests for Flight Memory Types
// Generic validation of Go bindings for Flight Memory Types

package integration

import (
	"encoding/json"
	"fmt"
	"testing"
	"time"

	memory "../../../bindings/go/enhanced"
)

// Generic utility functions for testing
type MemoryTestUtils struct{}

func (u MemoryTestUtils) CreateMemorySize(bytes uint64) memory.MemorySize {
	return memory.MemorySize{
		Bytes:         bytes,
		HumanReadable: u.FormatBytes(bytes),
	}
}

func (u MemoryTestUtils) FormatBytes(bytes uint64) string {
	units := []string{"B", "KB", "MB", "GB", "TB"}
	value := float64(bytes)
	unitIndex := 0

	for value >= 1024 && unitIndex < len(units)-1 {
		value /= 1024
		unitIndex++
	}

	return fmt.Sprintf("%.1f%s", value, units[unitIndex])
}

func (u MemoryTestUtils) CalculateUsagePercentage(snapshot memory.MemoryUsageSnapshot) float64 {
	return (float64(snapshot.Used.Bytes) / float64(snapshot.Total.Bytes)) * 100
}

func (u MemoryTestUtils) GetMemoryPressureLevel(percentage float64) memory.MemoryPressure {
	if percentage < 50 {
		return memory.MemoryPressureLow
	}
	if percentage < 80 {
		return memory.MemoryPressureMedium
	}
	if percentage < 95 {
		return memory.MemoryPressureHigh
	}
	return memory.MemoryPressureCritical
}

func (u MemoryTestUtils) CreateMemoryUpdate(sessionID string, snapshot memory.MemoryUsageSnapshot) memory.MemoryUpdate {
	return memory.MemoryUpdate{
		Type:      "memory_update",
		SessionID: sessionID,
		Snapshot:  snapshot,
		Timestamp: uint64(time.Now().Unix()),
	}
}

func (u MemoryTestUtils) CreateTestSnapshot(platform string, totalMB, usedMB uint64) memory.MemoryUsageSnapshot {
	total := u.CreateMemorySize(totalMB * 1024 * 1024)
	used := u.CreateMemorySize(usedMB * 1024 * 1024)
	available := u.CreateMemorySize((totalMB - usedMB) * 1024 * 1024)

	return memory.MemoryUsageSnapshot{
		Timestamp:          uint64(time.Now().Unix()),
		SessionID:          "test_session",
		Platform:           platform,
		Total:              total,
		Used:               used,
		Available:          available,
		FragmentationRatio: float32(usedMB) / float32(totalMB) * 0.1,
	}
}

func TestFlightMemoryTypes_GoIntegration(t *testing.T) {
	utils := MemoryTestUtils{}

	t.Run("Memory Size Creation and Formatting", func(t *testing.T) {
		// Dreamcast: 16MB
		dreamcast := utils.CreateMemorySize(16 * 1024 * 1024)
		if dreamcast.Bytes != uint64(16*1024*1024) {
			t.Errorf("Expected Dreamcast bytes %d, got %d", uint64(16*1024*1024), dreamcast.Bytes)
		}
		if dreamcast.HumanReadable != "16.0MB" {
			t.Errorf("Expected Dreamcast human readable '16.0MB', got '%s'", dreamcast.HumanReadable)
		}

		// PSP: 32MB
		psp := utils.CreateMemorySize(32 * 1024 * 1024)
		if psp.Bytes != uint64(32*1024*1024) {
			t.Errorf("Expected PSP bytes %d, got %d", uint64(32*1024*1024), psp.Bytes)
		}
		if psp.HumanReadable != "32.0MB" {
			t.Errorf("Expected PSP human readable '32.0MB', got '%s'", psp.HumanReadable)
		}

		// Vita: 512MB
		vita := utils.CreateMemorySize(512 * 1024 * 1024)
		if vita.Bytes != uint64(512*1024*1024) {
			t.Errorf("Expected Vita bytes %d, got %d", uint64(512*1024*1024), vita.Bytes)
		}
		if vita.HumanReadable != "512.0MB" {
			t.Errorf("Expected Vita human readable '512.0MB', got '%s'", vita.HumanReadable)
		}

		// Custom: 1GB
		custom := utils.CreateMemorySize(1024 * 1024 * 1024)
		if custom.Bytes != uint64(1024*1024*1024) {
			t.Errorf("Expected Custom bytes %d, got %d", uint64(1024*1024*1024), custom.Bytes)
		}
		if custom.HumanReadable != "1.0GB" {
			t.Errorf("Expected Custom human readable '1.0GB', got '%s'", custom.HumanReadable)
		}
	})

	t.Run("Byte Formatting Consistency", func(t *testing.T) {
		if utils.FormatBytes(1024) != "1.0KB" {
			t.Errorf("Expected '1.0KB', got '%s'", utils.FormatBytes(1024))
		}
		if utils.FormatBytes(1048576) != "1.0MB" {
			t.Errorf("Expected '1.0MB', got '%s'", utils.FormatBytes(1048576))
		}
		if utils.FormatBytes(1073741824) != "1.0GB" {
			t.Errorf("Expected '1.0GB', got '%s'", utils.FormatBytes(1073741824))
		}
		if utils.FormatBytes(1099511627776) != "1.0TB" {
			t.Errorf("Expected '1.0TB', got '%s'", utils.FormatBytes(1099511627776))
		}
	})

	t.Run("Memory Usage Calculations", func(t *testing.T) {
		snapshot := utils.CreateTestSnapshot("dreamcast", 16, 8)
		percentage := utils.CalculateUsagePercentage(snapshot)
		if percentage != 50.0 {
			t.Errorf("Expected usage percentage 50.0, got %f", percentage)
		}
	})

	t.Run("Platform Profile Support", func(t *testing.T) {
		platforms := []string{"dreamcast", "psp", "vita", "custom"}

		for _, platform := range platforms {
			snapshot := utils.CreateTestSnapshot(platform, 512, 256)
			if snapshot.Platform != platform {
				t.Errorf("Expected platform %s, got %s", platform, snapshot.Platform)
			}
			if snapshot.Total.Bytes != uint64(512*1024*1024) {
				t.Errorf("Expected total bytes %d, got %d", uint64(512*1024*1024), snapshot.Total.Bytes)
			}
			if snapshot.Used.Bytes != uint64(256*1024*1024) {
				t.Errorf("Expected used bytes %d, got %d", uint64(256*1024*1024), snapshot.Used.Bytes)
			}
		}
	})

	t.Run("JSON Serialization", func(t *testing.T) {
		snapshot := utils.CreateTestSnapshot("vita", 512, 256)
		update := utils.CreateMemoryUpdate("json_test", snapshot)

		// Test JSON marshaling
		jsonData, err := json.Marshal(update)
		if err != nil {
			t.Errorf("JSON marshaling failed: %v", err)
		}

		// Test JSON unmarshaling
		var unmarshaled memory.MemoryUpdate
		err = json.Unmarshal(jsonData, &unmarshaled)
		if err != nil {
			t.Errorf("JSON unmarshaling failed: %v", err)
		}

		if update.Type != unmarshaled.Type {
			t.Errorf("Expected type %s, got %s", update.Type, unmarshaled.Type)
		}
		if update.SessionID != unmarshaled.SessionID {
			t.Errorf("Expected session ID %s, got %s", update.SessionID, unmarshaled.SessionID)
		}
	})
}

// Benchmark tests for performance validation
func BenchmarkMemoryOperations(b *testing.B) {
	utils := MemoryTestUtils{}

	b.Run("MemorySize Creation", func(b *testing.B) {
		for i := 0; i < b.N; i++ {
			_ = utils.CreateMemorySize(uint64(i * 1024 * 1024))
		}
	})

	b.Run("Usage Percentage Calculation", func(b *testing.B) {
		snapshot := utils.CreateTestSnapshot("test", 1024, 512)
		b.ResetTimer()

		for i := 0; i < b.N; i++ {
			_ = utils.CalculateUsagePercentage(snapshot)
		}
	})

	b.Run("Memory Update Creation", func(b *testing.B) {
		snapshot := utils.CreateTestSnapshot("test", 512, 256)
		b.ResetTimer()

		for i := 0; i < b.N; i++ {
			_ = utils.CreateMemoryUpdate("bench_test", snapshot)
		}
	})

	b.Run("Bytes Formatting", func(b *testing.B) {
		for i := 0; i < b.N; i++ {
			_ = utils.FormatBytes(uint64(i * 1024 * 1024))
		}
	})

	b.Run("High Frequency Updates", func(b *testing.B) {
		sessionID := "high_freq_bench"
		b.ResetTimer()

		for i := 0; i < b.N; i++ {
			snapshot := utils.CreateTestSnapshot("vita", 512, uint64(i%512))
			_ = utils.CreateMemoryUpdate(sessionID, snapshot)
		}
	})

	b.Run("JSON Marshaling", func(b *testing.B) {
		snapshot := utils.CreateTestSnapshot("json_bench", 1024, 512)
		update := utils.CreateMemoryUpdate("json_test", snapshot)
		b.ResetTimer()

		for i := 0; i < b.N; i++ {
			_, _ = json.Marshal(update)
		}
	})
}

File: domains/flight-shared-types/tests/integration/typescript/memory-types.test.ts
/**
 * TypeScript Memory Types Integration Tests
 * Generic validation of TypeScript bindings for Flight Memory Types
 */

import {
  MemorySize,
  MemoryUsageSnapshot,
  MemoryPurpose,
  MemoryPressure,
  MemoryError,
  MemoryErrorCode,
  MemoryResult,
  SessionConfig,
  MemoryUpdate,
  PlatformProfile,
  MemoryAllocation,
  MemoryLimits,
  MemoryStats,
  MemoryTrend,
  TrendDirection
} from '../../../bindings/typescript/enhanced/src/types';

// Generic utility functions for testing
class MemoryTestUtils {
  static createMemorySize(bytes: bigint): MemorySize {
    const humanReadable = this.formatBytes(bytes);
    return { bytes, humanReadable };
  }

  static formatBytes(bytes: bigint): string {
    const units = ['B', 'KB', 'MB', 'GB', 'TB'];
    let value = Number(bytes);
    let unitIndex = 0;

    while (value >= 1024 && unitIndex < units.length - 1) {
      value /= 1024;
      unitIndex++;
    }

    return `${value.toFixed(1)}${units[unitIndex]}`;
  }

  static calculateUsagePercentage(snapshot: MemoryUsageSnapshot): number {
    return (Number(snapshot.used.bytes) / Number(snapshot.total.bytes)) * 100;
  }

  static getMemoryPressureLevel(percentage: number): MemoryPressure {
    if (percentage < 50) return MemoryPressure.Low;
    if (percentage < 80) return MemoryPressure.Medium;
    if (percentage < 95) return MemoryPressure.High;
    return MemoryPressure.Critical;
  }

  static createMemoryUpdate(sessionId: string, snapshot: MemoryUsageSnapshot): MemoryUpdate {
    return {
      type: 'memory_update',
      sessionId,
      snapshot,
      timestamp: Date.now()
    };
  }

  static createTestSnapshot(platform: string, totalMB: number, usedMB: number): MemoryUsageSnapshot {
    const total = this.createMemorySize(BigInt(totalMB * 1024 * 1024));
    const used = this.createMemorySize(BigInt(usedMB * 1024 * 1024));
    const available = this.createMemorySize(BigInt((totalMB - usedMB) * 1024 * 1024));
    
    return {
      timestamp: Date.now(),
      sessionId: 'test_session',
      platform,
      total,
      used,
      available,
      fragmentationRatio: usedMB / totalMB * 0.1
    };
  }
}

describe('Flight Memory Types - TypeScript Integration', () => {
  describe('Memory Size Creation and Formatting', () => {
    test('should create memory sizes for different platforms', () => {
      // Dreamcast: 16MB
      const dreamcast = MemoryTestUtils.createMemorySize(16n * 1024n * 1024n);
      expect(dreamcast.bytes).toBe(16n * 1024n * 1024n);
      expect(dreamcast.humanReadable).toBe('16.0MB');

      // PSP: 32MB
      const psp = MemoryTestUtils.createMemorySize(32n * 1024n * 1024n);
      expect(psp.bytes).toBe(32n * 1024n * 1024n);
      expect(psp.humanReadable).toBe('32.0MB');

      // Vita: 512MB
      const vita = MemoryTestUtils.createMemorySize(512n * 1024n * 1024n);
      expect(vita.bytes).toBe(512n * 1024n * 1024n);
      expect(vita.humanReadable).toBe('512.0MB');

      // Custom: 1GB
      const custom = MemoryTestUtils.createMemorySize(1024n * 1024n * 1024n);
      expect(custom.bytes).toBe(1024n * 1024n * 1024n);
      expect(custom.humanReadable).toBe('1.0GB');
    });

    test('should format bytes consistently across different sizes', () => {
      expect(MemoryTestUtils.formatBytes(1024n)).toBe('1.0KB');
      expect(MemoryTestUtils.formatBytes(1048576n)).toBe('1.0MB');
      expect(MemoryTestUtils.formatBytes(1073741824n)).toBe('1.0GB');
      expect(MemoryTestUtils.formatBytes(1099511627776n)).toBe('1.0TB');
    });
  });

  describe('Memory Usage Calculations', () => {
    test('should calculate usage percentages correctly', () => {
      const snapshot = MemoryTestUtils.createTestSnapshot('dreamcast', 16, 8);
      const percentage = MemoryTestUtils.calculateUsagePercentage(snapshot);
      expect(percentage).toBe(50.0);
    });

    test('should determine memory pressure levels accurately', () => {
      expect(MemoryTestUtils.getMemoryPressureLevel(30)).toBe(MemoryPressure.Low);
      expect(MemoryTestUtils.getMemoryPressureLevel(60)).toBe(MemoryPressure.Medium);
      expect(MemoryTestUtils.getMemoryPressureLevel(90)).toBe(MemoryPressure.High);
      expect(MemoryTestUtils.getMemoryPressureLevel(98)).toBe(MemoryPressure.Critical);
    });
  });

  describe('Platform Profile Support', () => {
    test('should support all platform profiles', () => {
      const platforms = ['dreamcast', 'psp', 'vita', 'custom'];
      
      platforms.forEach(platform => {
        const snapshot = MemoryTestUtils.createTestSnapshot(platform, 512, 256);
        expect(snapshot.platform).toBe(platform);
        expect(snapshot.total.bytes).toBe(512n * 1024n * 1024n);
        expect(snapshot.used.bytes).toBe(256n * 1024n * 1024n);
      });
    });

    test('should handle custom platform configurations', () => {
      const customSnapshot = MemoryTestUtils.createTestSnapshot('custom-2gb', 2048, 1024);
      expect(customSnapshot.platform).toBe('custom-2gb');
      expect(customSnapshot.total.humanReadable).toBe('2.0GB');
      expect(customSnapshot.used.humanReadable).toBe('1.0GB');
    });
  });

  describe('Memory Allocation Tracking', () => {
    test('should create memory allocations for different purposes', () => {
      const allocations: MemoryAllocation[] = [
        {
          id: 'alloc_1',
          sessionId: 'session_1',
          size: MemoryTestUtils.createMemorySize(1024n * 1024n),
          purpose: MemoryPurpose.VmHeap,
          allocatedAt: Date.now()
        },
        {
          id: 'alloc_2', 
          sessionId: 'session_1',
          size: MemoryTestUtils.createMemorySize(512n * 1024n),
          purpose: MemoryPurpose.AssetCache,
          allocatedAt: Date.now()
        }
      ];

      expect(allocations[0].purpose).toBe(MemoryPurpose.VmHeap);
      expect(allocations[0].size.humanReadable).toBe('1.0MB');
      expect(allocations[1].purpose).toBe(MemoryPurpose.AssetCache);
      expect(allocations[1].size.humanReadable).toBe('512.0KB');
    });
  });

  describe('Error Handling', () => {
    test('should create and handle memory errors properly', () => {
      const error: MemoryError = {
        code: MemoryErrorCode.LimitExceeded,
        message: 'Memory limit exceeded',
        details: 'Attempted allocation larger than available memory',
        timestamp: Date.now()
      };

      expect(error.code).toBe(MemoryErrorCode.LimitExceeded);
      expect(error.message).toContain('Memory limit exceeded');
      expect(error.details).toContain('Attempted allocation');
    });

    test('should handle memory result types correctly', () => {
      // Test successful result
      const successResult: MemoryResult<boolean> = {
        tag: 'ok',
        val: true
      };

      expect(successResult.tag).toBe('ok');
      expect(successResult.val).toBe(true);

      // Test error result
      const errorResult: MemoryResult<boolean> = {
        tag: 'err',
        val: {
          code: MemoryErrorCode.AllocationFailed,
          message: 'Failed to allocate memory',
          timestamp: Date.now()
        }
      };

      expect(errorResult.tag).toBe('err');
      expect((errorResult.val as MemoryError).code).toBe(MemoryErrorCode.AllocationFailed);
    });
  });

  describe('Real-time Updates and Events', () => {
    test('should create memory updates for streaming', () => {
      const snapshot = MemoryTestUtils.createTestSnapshot('psp', 32, 16);
      const update = MemoryTestUtils.createMemoryUpdate('session_123', snapshot);

      expect(update.type).toBe('memory_update');
      expect(update.sessionId).toBe('session_123');
      expect(update.snapshot.platform).toBe('psp');
      expect(typeof update.timestamp).toBe('number');
    });

    test('should handle high-frequency updates efficiently', () => {
      const sessionId = 'high_freq_session';
      const updates: MemoryUpdate[] = [];

      // Simulate 1000 rapid updates
      const startTime = performance.now();
      for (let i = 0; i < 1000; i++) {
        const snapshot = MemoryTestUtils.createTestSnapshot('vita', 512, i % 512);
        const update = MemoryTestUtils.createMemoryUpdate(sessionId, snapshot);
        updates.push(update);
      }
      const endTime = performance.now();

      const totalTime = endTime - startTime;
      const avgTimePerUpdate = totalTime / 1000;

      // Should be very fast - under 0.1ms per update for <1% overhead
      expect(avgTimePerUpdate).toBeLessThan(0.1);
      expect(updates.length).toBe(1000);
      expect(updates[0].sessionId).toBe(sessionId);
    });
  });

  describe('Session Management', () => {
    test('should handle multiple concurrent sessions', () => {
      const sessions: SessionConfig[] = [
        {
          sessionId: 'session_1',
          userId: 'user_1',
          platform: 'dreamcast'
        },
        {
          sessionId: 'session_2',
          userId: 'user_2', 
          platform: 'psp'
        },
        {
          sessionId: 'session_3',
          userId: 'user_3',
          platform: 'vita'
        }
      ];

      sessions.forEach((session, index) => {
        expect(session.sessionId).toBe(`session_${index + 1}`);
        expect(session.userId).toBe(`user_${index + 1}`);
        expect(['dreamcast', 'psp', 'vita']).toContain(session.platform);
      });
    });
  });

  describe('Memory Analytics and Trends', () => {
    test('should track memory statistics correctly', () => {
      const stats: MemoryStats = {
        totalAllocations: 10,
        activeAllocations: 7,
        peakMemory: MemoryTestUtils.createMemorySize(256n * 1024n * 1024n),
        currentMemory: MemoryTestUtils.createMemorySize(128n * 1024n * 1024n),
        averageAllocationSize: MemoryTestUtils.createMemorySize(16n * 1024n * 1024n),
        usageByPurpose: [
          [MemoryPurpose.VmHeap, MemoryTestUtils.createMemorySize(64n * 1024n * 1024n)],
          [MemoryPurpose.AssetCache, MemoryTestUtils.createMemorySize(32n * 1024n * 1024n)]
        ],
        efficiencyRatio: 0.85
      };

      expect(stats.totalAllocations).toBe(10);
      expect(stats.activeAllocations).toBe(7);
      expect(stats.peakMemory.humanReadable).toBe('256.0MB');
      expect(stats.currentMemory.humanReadable).toBe('128.0MB');
      expect(stats.efficiencyRatio).toBe(0.85);
    });

    test('should analyze memory trends', () => {
      const snapshots = [
        MemoryTestUtils.createTestSnapshot('vita', 512, 100),
        MemoryTestUtils.createTestSnapshot('vita', 512, 200),
        MemoryTestUtils.createTestSnapshot('vita', 512, 300)
      ];

      const trend: MemoryTrend = {
        snapshots,
        trendDirection: TrendDirection.Increasing,
        predictedPeak: MemoryTestUtils.createMemorySize(400n * 1024n * 1024n)
      };

      expect(trend.snapshots.length).toBe(3);
      expect(trend.trendDirection).toBe(TrendDirection.Increasing);
      expect(trend.predictedPeak?.humanReadable).toBe('400.0MB');
    });
  });
});
